### PR TITLE
Hackday: Optimise site for accessibility, offline support and performance

### PR DESCRIPTION
**Pull request by barrymcgee**
## Done

- Fix accessibility issues
- Lazy load all images on page load
- Compress file of all images
- Add distinct meta descriptions for each page
- Add service worker for offline support
- Add app icons to make site installable as a PWA

## QA 

I used Lighthouse within Chrome Dev Tools to identify issues we could improve. 

### Benchmark on production

screen shot 2018-07-26 at 11 20 21

### Running on demo site with this changeset

screen shot 2018-07-26 at 12 00 22

There are still other performance improvements that can be made such as;

- inlining critical CSS
- removing unused CSS
- removing Sourcemaps from production CSS (I think this is a bug upstream with `./run`)

### Checking offline support

- Run the demo site: https://vfio-28938.firebaseapp.com/
- Disable your internet connection
- Refresh the page
- The site should still load 💥

### CODE REVIEW

**Comment by barrymcgee** [comment]
Demo started at: https://vfio-28938.firebaseapp.com/

**Comment by anthonydillon** [comment]
I see different content on this branch when running it locally:
screenshot_2018-07-31 vanilla framework

File: sw.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        fetch(request).then(function(response) {
+          cache.put(request, response);
+        });
+      });
+
+      return cache.addAll([
+        '/',
+        '/index.html',
+        '/index.html?homescreen=1',
+        '/?homescreen=1',
```
Comment (anthonydillon) [inline sw.js]: What are these for?

File: sw.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  '/v1/01c0e480-landscape_black-orange_hex.svg',
+  '/v1/9d12b4c4-maas_black-orange_hex.svg',
+  '/v1/ffa38040-juju_black-orange_hex.svg',
+  '/v1/41e2722c-core_logo.svg',
+  '/v1/3f0091ee-conjure-up-logo-black.svg',
+  '/v1/ea4e7803-snapcraft_green-red_su_hex_cropped.svg',
+  '/v1/e0d78513-cloud-init-logo-updated.svg',
+  '/v1/f53c4e01-icon-slack.svg',
+  '/v1/7eae50ca-icon-twitter.svg',
+  '/v1/f307a122-icon-github.svg'
```
Comment (anthonydillon) [inline sw.js]: These are likely to become dated if the content on the site changes. Is there a way to make the images automatically cache?

File: manifest.json [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+      "src": "app-icons/firefox/firefox-general-48-48.png",
+      "sizes": "48x48"
+    },
+    {
+      "src": "app-icons/firefox/firefox-general-32-32.png",
+      "sizes": "32x32"
+    },
+    {
+      "src": "app-icons/firefox/firefox-general-16-16.png",
+      "sizes": "16x16"
```
Comment (anthonydillon) [inline manifest.json]: Do we really need all these images?

Comment (barrymcgee) [reply manifest.json]: They came from an icon generator which aimed to cover all bases. If you specify which icons you want to keep I can remove the rest?

File: sw.js (hunk 1 above)
Comment (barrymcgee) [reply sw.js]: These are the URLs you return to after you install the app on your home screen etc. - they differ across different platforms which is why you need two variations.

File: sw.js (hunk 2 above)
Comment (barrymcgee) [reply sw.js]: @steverydz worked on this part but from my brief investigation, I don't think so - even the example on MDN hardcodes them:

https://developer.mozilla.org/en-US/docs/Web/API/Cache/addAll

**Comment by barrymcgee** [comment]
@anthonydillon What content is different?

**Comment by barrymcgee** [comment]
A bug I've just found is that when offline, all other pages except for the index page return the 404 page 🧐

I've played around but can't resolve it - any ideas @steverydz?

**Comment by steverydz** [comment]
@barrymcgee It works for me offline as long as the .html suffix is there. Without that I get 404 online and offline.

**Comment by kwm14** [comment]
@barrymcgee any update on this PR? 😃

**Comment by barrymcgee** [comment]
@kwm14 We need to resolve the issue above - can your squad pick it up this sprint?

**Comment by kwm14** [comment]
@barrymcgee we don't have capacity this iteration. But @deadlight or @anthonydillon will take a look at this next iteration under _Small tasks_ 👍

**Comment by anthonydillon** [comment]
@steverydz @barrymcgee I think we should merge this. I have rebased on master and its miles better:

## Before:
screenshot from 2018-12-04 23-52-48

## After:
screenshot from 2018-12-04 23-52-53

**Comment by anthonydillon** [comment]
The drops in best practises and SEO are due to differences on the demo server from prod

**Comment by barrymcgee** [comment]
@anthonydillon The offline bug persists but as you say, it's a big improvement regardless so should be merged.

**Comment by kwm14** [comment]
@anthonydillon @barrymcgee @deadlight if this fix seems ok can someone update labels and merge please 👍 😄 💯

**Comment by barrymcgee** [comment]
I've just rebased and resolved the conflict :shipit:

**Pull request merged**